### PR TITLE
Add list box #121

### PR DESCRIPTION
**Pull request by stonko1994**
Added a new UI Element discussed [here](https://github.com/bitmovin/bitmovin-player-ui/pull/121)

### CODE REVIEW

File: src/ts/components/listbox.ts [hunk 1]
```diff
@@ -0,0 +1,173 @@
+import {UIInstanceManager} from '../uimanager';
+import {Button, ButtonConfig} from './button';
+import {ListSelector, ListSelectorConfig} from './listselector';
+import {DOM} from '../dom';
+import {Timeout} from '../timeout';
+
+export interface ListBoxConfig extends ListSelectorConfig {
```
Comment (protyposis) [inline src/ts/components/listbox.ts]: The `ListBox` should be a component analogous to `SelectBox`, i.e. it would be used for example as a child inside a `SettingsPanel`. A `SettingsPanel` can contain one or many `ListBox` instances. A `SelectBox` also does not have a cointainer or cares about hiding, that's the business of whatever `Container` the `SelectBox` or `ListBox` night be put into. This is also important in regards to paging (#119), where a `SettingsPanel` will be able to contain multiple `SettingsPanel`s that can be navigated.

It seems what you have created here is a hybrid of a `SettingsPanel` and a `ListBox`. This goes against the design principle of component reuse and composition.

File: src/ts/components/listboxtogglebutton.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
+import {ToggleButton, ToggleButtonConfig} from './togglebutton';
+import {ListBox} from './listbox';
+import {UIInstanceManager} from '../uimanager';
+
+export interface ListBoxToggleButtonConfig extends ToggleButtonConfig {
```
Comment (protyposis) [inline src/ts/components/listboxtogglebutton.ts]: Continuing the comment from above, if the `ListBox` is a child of a `SettingsPanel`, there's also no need for this component but the already existing `SettingsToggleButton` will be used. Of course there can be a `SubtitleSettingsToggleButton` if a button with extended functionality specifically for the subtitle settings panel is required.

File: src/ts/components/listbox.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+ *       <div class='ui-listboxitem'>
+ *         <button class='ui-listboxitem-button'>label</button>
+ *       </div>
+ *     </div>
+ *     ...
+ *   </div
+ * </code>
+ */
+export class ListBox extends ListSelector<ListSelectorConfig> {
+  private listBoxElement: DOM;
```
Comment (protyposis) [inline src/ts/components/listbox.ts]: This field is unnecessary. The `DOM` object is already stored in `Component` and can be retrieved with `getDomElement()`

File: src/ts/components/listbox.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+    this.listBoxElement = listBoxElement;
+    this.updateDomItems();
+
+    return listBoxElement;
+  }
+
+  protected updateDomItems(selectedValue: string = null) {
+    // Delete all children
+    this.listBoxElement.empty();
```
Comment (protyposis) [inline src/ts/components/listbox.ts]: Use `this.getDomElement()` instead of `tihs.listBoxElement`

File: src/ts/components/listbox.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+    // Add updated children
+    for (let item of this.items) {
+      let rowElement = new DOM('div', {
+        'class': this.prefixCss(ListBox.LIST_BOX_ITEM_WRAPPER_CLASS),
```
Comment (protyposis) [inline src/ts/components/listbox.ts]: What is this wrapper necessary for? Why can't you directly work with the `ListBoxItemButton`?

File: src/scss/skin-modern/components/_listbox.scss [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        background-color: transparentize($color-highlight, .15);
+      }
+
+      &:last-child {
+        border-bottom: 0;
+      }
+    }
+  }
+}
+
```
Comment (protyposis) [inline src/scss/skin-modern/components/_listbox.scss]: You are missing the actual class declaration here:

```scss
.#{$prefix}-ui-listbox {
  @extend %ui-listbox;
}
```

Without that, a basic `ListBox` would not be styled.

File: src/scss/skin-modern/components/_subtitlesettingspanel.scss [hunk 7]
```diff
@@ -0,0 +1,53 @@
+@import '../variables';
+@import '../mixins';
+
+.#{$prefix}-ui-subtitles-settings-panel {
```
Comment (protyposis) [inline src/scss/skin-modern/components/_subtitlesettingspanel.scss]: What is this file necessary for? Why is the styling not covered by `_settingspanel.scss`?

Basically, for every SCSS in the `components` folder, there must be a `Component`, which in this case is violated. You should generify the `SettingsPanel` so it covers also the subtitle settings.

File: src/scss/skin-modern/components/_subtitlesettingspaneltogglebutton.scss [hunk 8]
```diff
@@ -0,0 +1,11 @@
+// demo for extracted audio tracks and subtitle settings from the settings panel direct into the controlBar
+@import '../variables';
```
Comment (protyposis) [inline src/scss/skin-modern/components/_subtitlesettingspaneltogglebutton.scss]: Again there is no equivalent `Component` class. In this case, you can create a `SubtitleSettingsPanelToggleButton extends SettingsPanelToggleButton`.

File: src/ts/components/subtitleslistbox.ts [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+  constructor(config: ListSelectorConfig = {}) {
+    super(config);
+    this.config = this.mergeConfig(config, {
+      cssClass: 'ui-subtitles-listbox',
+    }, this.config);
+  }
+
+  configure(player: bitmovin.PlayerAPI, uimanager: UIInstanceManager): void {
+    super.configure(player, uimanager);
+    let selectCurrentSubtitle = () => {
```
Comment (protyposis) [inline src/ts/components/subtitleslistbox.ts]: This is basically the same code as in the `SubtitleSelectBox`, so the code should be shared via a helper/utility class.

File: src/ts/components/subtitleslistbox.ts [hunk 10]
```diff
@@ -0,0 +1,55 @@
+import {ListBox} from './listbox';
+import {ListSelectorConfig} from './listselector';
+import {UIInstanceManager} from '../uimanager';
+
+/**
+ * A element that is similar to a select box where the user can select a subtitle
+ */
+export class SubtitlesListBox extends ListBox {
```
Comment (protyposis) [inline src/ts/components/subtitleslistbox.ts]: The name should be in singular, i.e. `SubtitleListBox`, Analogous to `SubtitleSelectBox` (filename too)

File: src/scss/skin-modern/components/_subtitlesettingspaneltogglebutton.scss (hunk 8 above)
Comment (stonko1994) [reply src/scss/skin-modern/components/_subtitlesettingspaneltogglebutton.scss]: I had one but i was not sure because the only different to the `SettingsPanelToggleButton` was a different `cssClass`.

File: src/scss/skin-modern/components/_subtitlesettingspanel.scss (hunk 7 above)
Comment (stonko1994) [reply src/scss/skin-modern/components/_subtitlesettingspanel.scss]: It's needed for the different positioning and the little arrow to its toggle button.

File: src/ts/components/listbox.ts (hunk 4 above)
Comment (stonko1994) [reply src/ts/components/listbox.ts]: This will result in an infinite loop because the `updateDomItems` are called from the `getDomElement` function to handle initial state configuration.

Similar to the `SelectBox`

File: src/scss/skin-modern/components/_subtitlesettingspaneltogglebutton.scss (hunk 8 above)
Comment (protyposis) [reply src/scss/skin-modern/components/_subtitlesettingspaneltogglebutton.scss]: Yes both is not really ideal. Ok then please leave the file in its place but extend the `ui-settingstogglebutton` instead of `ui-button`.

File: src/scss/skin-modern/components/_subtitlesettingspanel.scss (hunk 7 above)
Comment (protyposis) [reply src/scss/skin-modern/components/_subtitlesettingspanel.scss]: Ok but why is that whole file necessary to just change the positioning? Shouldn't that be much simpler?

File: src/ts/components/listbox.ts (hunk 4 above)
Comment (protyposis) [reply src/ts/components/listbox.ts]: Ok then it's fine :)

File: src/scss/skin-modern/components/_subtitlesettingspanel.scss (hunk 7 above)
Comment (stonko1994) [reply src/scss/skin-modern/components/_subtitlesettingspanel.scss]: Should be. We can consider using something like a wrapper over the `SettingsToggleButton` and the `SettingsPanel` to 'group' then together. Than the positioning and the little arrow should be easier.

But i would say to do this in a separate issue and then first extract the `Panel` and introduce it also for the `SettingsPanel`

File: src/ts/components/subtitleslistbox.ts (hunk 9 above)
Comment (stonko1994) [reply src/ts/components/subtitleslistbox.ts]: Hm what about a `SubtitleSwitchHandler` and pass the `bitmovin.PlayerAPI` and a subclass of `ListSelector` and handle all events where something is to update (Player Events & ListSelector Events) within this class?

File: src/scss/skin-modern/components/_subtitlesettingspanel.scss (hunk 7 above)
Comment (protyposis) [reply src/scss/skin-modern/components/_subtitlesettingspanel.scss]: Even without a wrapper, I don't think it's necessary to replicate that much of the `_settingspanel.scss`. This file also should not contain styles for a `ListBox`.

File: src/ts/components/subtitleslistbox.ts (hunk 9 above)
Comment (protyposis) [reply src/ts/components/subtitleslistbox.ts]: Sounds good!

File: src/scss/skin-modern/components/_settingstogglebutton.scss [hunk 11]
```diff
@@ -1,15 +1,25 @@
 @import '../variables';
 @import '../mixins';
 
-.#{$prefix}-ui-settingstogglebutton {
+%ui-togglebutton {
```
Comment (protyposis) [inline src/scss/skin-modern/components/_settingstogglebutton.scss]: Should be named `ui-settingstogglebutton` (like the file)

File: src/scss/skin-modern/components/_subtitlesettingspaneltogglebutton.scss [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+
+.#{$prefix}-ui-subtitles-settings-toggle-button {
```
Comment (protyposis) [inline src/scss/skin-modern/components/_subtitlesettingspaneltogglebutton.scss]: `subtitle` (singular)

File: src/ts/subtitleutils.ts [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * Helper class to handle all subtitle related events
+ *
+ * This class listens to player events as well as the `ListSelector` event if selection changed
+ */
+export class SubtitleSwitchHandler {
+  private player: bitmovin.PlayerAPI;
+  private listElement: ListSelector<ListSelectorConfig>;
+
+  constructor(player: bitmovin.PlayerAPI, element: SubtitleListBox | SubtitleSelectBox) {
```
Comment (protyposis) [inline src/ts/subtitleutils.ts]: Can't the type be `element: ListSelector`?

File: src/ts/subtitleutils.ts [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+    this.player = player;
+    this.listElement = element;
+
+    this.bindSelectionEvent();
+    this.bindPlayerEvents();
+    const callback = this.buildUpdateSubtitlesCallback();
```
Comment (protyposis) [inline src/ts/subtitleutils.ts]: Why do you need a `build` method instead of calling it directly? (Applies to all usages of this method)

File: src/ts/subtitleutils.ts [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+import {SubtitleListBox} from './components/subtitlelistbox';
+import {SubtitleSelectBox} from './components/subtitleselectbox';
+
+/**
+ * Helper class to handle all subtitle related events
+ *
+ * This class listens to player events as well as the `ListSelector` event if selection changed
+ */
+export class SubtitleSwitchHandler {
+  private player: bitmovin.PlayerAPI;
```
Comment (protyposis) [inline src/ts/subtitleutils.ts]: Please always add a blank line between the `class` declaration and the first field/constructor/method/etc.

File: src/ts/subtitleutils.ts [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+    this.player = player;
+    this.listElement = element;
+
+    this.bindSelectionEvent();
+    this.bindPlayerEvents();
+    const callback = this.buildUpdateSubtitlesCallback();
+    callback();
+  }
+
+  bindSelectionEvent(): void {
```
Comment (protyposis) [inline src/ts/subtitleutils.ts]: Make all methods private

File: src/scss/skin-modern/components/_subtitlesettingspanel.scss [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+  @extend %ui-settings-panel;
+
+  right: 3em;
+  width: 7em;
+
+  // bottom arrow from http://www.cssarrowplease.com/
+  &::after {
```
Comment (protyposis) [inline src/scss/skin-modern/components/_subtitlesettingspanel.scss]: How about we remove that arrow completely and with that get rid of the need for correct placement? Looks childish anyway. What do you think?

File: src/ts/components/subtitlelistbox.ts [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+    super(config);
+    this.config = this.mergeConfig(config, {
+      cssClass: 'ui-subtitles-listbox',
+    }, this.config);
+  }
+
+  configure(player: bitmovin.PlayerAPI, uimanager: UIInstanceManager): void {
+    super.configure(player, uimanager);
+
+    new SubtitleSwitchHandler(player, this);
```
Comment (protyposis) [inline src/ts/components/subtitlelistbox.ts]: very nice :)

File: src/ts/uimanager.ts [hunk 19]
```diff
@@ -1061,11 +1072,9 @@ export namespace UIManager.Demo.Factory {
...
-            new AudioTrackSelectBox({
-              cssClasses: ['my-audio-track-selection-demo'],
-            }),
-            new SubtitleSelectBox({
-              cssClasses: ['my-subtitle-selection-demo'],
+            new SettingsToggleButton({
+              settingsPanel: subtitleSettingsPanel,
+              cssClass: 'ui-subtitles-settings-toggle-button',
             }),
             new SettingsToggleButton({ settingsPanel: settingsPanel }),
```
Comment (protyposis) [inline src/ts/uimanager.ts]: You need to hide all visible settingspanels when another panel is opened. Currently we can have both visible at the same time. You can achieve this via the `UIInstanceManager.onComponentShow` event:

```ts
// Keep track of shown SettingsPanels
const shownSettingsPanels: SettingsPanel[] = [];
uiManager.onComponentShow.subscribe((sender) => {
  if (sender instanceof SettingsPanel) {
    shownSettingsPanels.push(sender);
    sender.onComponentHide.subscribeOnce(() => ArrayUtils.remove(shownSettingsPanels, sender));
  }
});

// Hide all open SettingsPanels before opening this button's panel
// (We need to iterate a copy because hiding them will automatically remove themselves from the array
// due to the subscribeOnce above)
shownSettingsPanels.slice().forEach(settingsPanel => settingsPanel.hide());

mySettingsPanel.show();
```

(this code is not tested but should give you the idea)

File: src/ts/components/listbox.ts [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+        key: item.key,
+        text: item.label,
+      });
+
+      itemButton.onClick.subscribe((sender) => {
+        this.handleSelectionChange(<ListBoxItemButton>sender);
+      });
+
+      const selectedItemClass = this.prefixCss(ListBox.SELECTED_LIST_BOX_ITEM_CLASS);
+      const itemElement = itemButton.getDomElement();
```
Comment (protyposis) [inline src/ts/components/listbox.ts]: This is actually a hack. To do it right, the `ListSelector` must be converted to extend `Container` and the `SelectBox` needs to be adjusted to that. Currently you are managing `Button` instances outside of the component tree. This might have side effects. But lets let that be for the moment.

File: src/ts/subtitleutils.ts (hunk 14 above)
Comment (stonko1994) [reply src/ts/subtitleutils.ts]: because when passing the method direct to the player event callbacks we are loosing the lexical scope.

Comment (protyposis) [reply src/ts/subtitleutils.ts]: You mean the `this` context (binding)?

File: src/scss/skin-modern/components/_subtitlesettingspanel.scss (hunk 17 above)
Comment (stonko1994) [reply src/scss/skin-modern/components/_subtitlesettingspanel.scss]: hm yes we can do it. I'm fine with that.

File: src/ts/subtitleutils.ts (hunk 14 above)
Comment (stonko1994) [reply src/ts/subtitleutils.ts]: yes

File: CHANGELOG.md [hunk 21]
```diff
@@ -7,6 +7,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 ## [develop]
 
 ### Added
+- New `ListBox` UI element to display multiple selectable items
```
Comment (protyposis) [inline CHANGELOG.md]: `New` is redundant. It's implicitly new as it's in the `Added` section. Just writing `` `ListBox` UI element`` is sufficient.

Also add an entry for the demo, e.g.

```
### Added
- Demo section in playground with separate `SettingsPanel`s for subtitle and audio track selection via `ListBox`
```

File: CHANGELOG.md [hunk 22]
```diff
@@ -7,6 +7,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
...
 - Icon for `Subtitles`
```
Comment (protyposis) [inline CHANGELOG.md]: `Subtitles` and `AudioTracks` are not componentos or other classes within the UI framework so they don't belong into `code` markup

File: src/scss/skin-modern/components/_audiotracklistbox.scss [hunk 23]
```diff
@@ -0,0 +1,6 @@
+@import '../variables';
+@import '../mixins';
+
+.#{$prefix}-ui-audiotrack-listbox {
```
Comment (protyposis) [inline src/scss/skin-modern/components/_audiotracklistbox.scss]: This file is unnecessary. There is also no `_audiotrackselectbox.scss`.

File: src/scss/skin-modern/components/_audiotracksettingspanel.scss [hunk 24]
```diff
@@ -0,0 +1,29 @@
+@import '../variables';
+@import '../mixins';
+
+.#{$prefix}-ui-audiotrack-settings-panel {
+  @extend %ui-settings-panel;
+
+  right: 5em;
```
Comment (protyposis) [inline src/scss/skin-modern/components/_audiotracksettingspanel.scss]: The settings panels should all be rendered in the same place by default.

File: src/scss/skin-modern/components/_subtitlesettingspanel.scss [hunk 25]
```diff
@@ -0,0 +1,29 @@
+@import '../variables';
+@import '../mixins';
+
+.#{$prefix}-ui-subtitle-settings-panel {
+  @extend %ui-settings-panel;
+
+  right: 3em;
```
Comment (protyposis) [inline src/scss/skin-modern/components/_subtitlesettingspanel.scss]: Same here. Use default positioning.

File: src/ts/main.ts [hunk 26]
```diff
@@ -131,6 +132,7 @@ const playerui = {
   HugePlaybackToggleButton,
   HugeReplayButton,
   Label,
+  ListBox,
```
Comment (protyposis) [inline src/ts/main.ts]: Also export the audio track and subtitle list boxes

File: src/ts/subtitleutils.ts (hunk 14 above)
Comment (protyposis) [reply src/ts/subtitleutils.ts]: The `build*` methods need to be removed. Write the methods directly as arrow functions to get rid of the function return value.

File: src/ts/demofactory.ts [hunk 27]
```diff
@@ -90,11 +110,13 @@ export namespace DemoFactory {
             new AirPlayToggleButton(),
             new CastToggleButton(),
             new VRToggleButton(),
-            new AudioTrackSelectBox({
-              cssClasses: ['my-audio-track-selection-demo'],
+            new SettingsToggleButton({
+              settingsPanel: audioTrackSettingsPanel,
+              cssClass: 'ui-audiotrack-setting-toggle-button',
```
Comment (protyposis) [inline src/ts/demofactory.ts]: The `cssClass` of the `SettingsToggleButton` is `ui-settingstogglebutton`, so this should be something along `ui-audiotracksettingstogglebutton` (the pattern is always to extend the parent's name).

File: src/ts/demofactory.ts [hunk 28]
```diff
@@ -90,11 +110,13 @@ export namespace DemoFactory {
...
             }),
-            new SubtitleSelectBox({
-              cssClasses: ['my-subtitle-selection-demo'],
+            new SettingsToggleButton({
+              settingsPanel: subtitleSettingsPanel,
+              cssClass: 'ui-subtitle-setting-toggle-button',
```
Comment (protyposis) [inline src/ts/demofactory.ts]: As above.

File: src/scss/skin-modern/components/_audiotracklistbox.scss (hunk 23 above)
Comment (protyposis) [reply src/scss/skin-modern/components/_audiotracklistbox.scss]: Same goes for `_subtitlelistbox.scss`

File: src/scss/skin-modern/components/_audiotracksettingspanel.scss [hunk 29]
```diff
@@ -0,0 +1,28 @@
+@import '../variables';
+@import '../mixins';
+
+.#{$prefix}-ui-audiotrack-settings-panel {
```
[the remark this answers is not in the document]
Comment (stonko1994) [reply src/scss/skin-modern/components/_audiotracksettingspanel.scss]: To improve styling for a `settings-panel` when there is a `list-box` in it. Same for `_subtitlesettingspanel.scss` but can be combined.

[the remark this answers is not in the document]
Comment (protyposis) [reply src/scss/skin-modern/components/_audiotracksettingspanel.scss]: And it's not possible to tweak the base style?

File: src/ts/components/listbox.ts [hunk 30]
```diff
@@ -0,0 +1,117 @@
...
+ */
+export class ListBox extends ListSelector<ListSelectorConfig> {
+  private listBoxElement: DOM;
+
+  private static readonly SELECTED_LIST_BOX_ITEM_CLASS = 'ui-listboxitem-button-selected';
+
+  constructor(config: ListSelectorConfig = {}) {
+    super(config);
+
+    this.config = this.mergeConfig(config, {
```
[the remark this answers is not in the document]
Comment (stonko1994) [reply src/ts/components/listbox.ts]: Done in `subtitlelistbox.ts`.

File: src/scss/skin-modern/components/_audiotracksettingspanel.scss (hunk 29 above)
[the remark this answers is not in the document]
Comment (stonko1994) [reply src/scss/skin-modern/components/_audiotracksettingspanel.scss]: Found a solution for the label but not for the margin / padding.
